crypto/stack/stack.c

```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 
     if ((ret = OPENSSL_zalloc(sizeof(*ret))) == NULL)
         goto err;
-    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * MIN_NODES)) == NULL)
+    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * min_nodes)) == NULL)
         goto err;
     ret->comp = c;
-    ret->num_alloc = MIN_NODES;
+    ret->num_alloc = min_nodes;
     return (ret);
 
  err:
     OPENSSL_free(ret);
     return (NULL);
 }
 
-int OPENSSL_sk_insert(OPENSSL_STACK *st, const void *data, int loc)
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth faction is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use 3/2 = 1.5 for simplicty of calculation and overflow checking.
+ * Another option 8/5 = 1.6 allows for slightly faster growth, although safe
+ * computation is more difficult.
+ *
+ * The limit to avoid overflow is spot on.  The modulo three correction term
+ * ensures that the limit is the largest number than can be expanded by the
+ * growth factor without exceeding the hard limit.
+ */
+static ossl_inline int compute_growth(int target, int current)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
-        return 0;
+    const int limit = (max_nodes / 3) * 2 + max_nodes % 3;
+
+    while (current <= target) {
+        /* Check to see if we're at the hard limit */
+        if (current >= max_nodes)
+            return 0;
+
+        /* Expand the size by a factor of 3/2 if it is within range */
+        current = current < limit ? current + current / 2 : max_nodes;
     }
+    return current;
+}
 
-    if (st->num_alloc <= (size_t)(st->num + 1)) {
-        size_t doub_num_alloc = st->num_alloc * 2;
-        const char **tmpdata;
+static int sk_reserve(OPENSSL_STACK *st, int n, int exact)
+{
+    const void **tmpdata;
+    int num_alloc;
 
-        /* Overflow checks */
-        if (doub_num_alloc < st->num_alloc)
-            return 0;
+    /* Check to see the reservation isn't exceeding the hard limit */
+    if (n > max_nodes - st->num)
+        return 0;
 
-        /* Avoid overflow due to multiplication by sizeof(char *) */
-        if (doub_num_alloc > SIZE_MAX / sizeof(char *))
+    /* Figure out the new size */
+    num_alloc = st->num + n;
+    if (!exact) {
+        if (num_alloc <= st->num_alloc)
+            return 1;
```

Comment:
Speaking of this condition. Wouldn't it be appropriate to check for equality in exact case? [For reference, my first reaction was to suggest to take out of `if (!exact)`, but then I realized that it would prevent freeing some memory, which might be desirable.]

Reply:
Does the exact + equality fall out with the later `num_alloc == st->num` test? a few lines below?

Reply:
Oh! But no, it doesn't cut it. Because it's different condition, and one can even wonder what's up with that. What does size of array have to do number of initialized elements? I mean I'm struggling with question why look at st->num at all. At least at reservation time it's likely to be 0, isn't it?

Reply:
It the context one can wonder if it makes sense to postpone allocation of st->data till first reservation or insertion. On can also wonder if it's appropriate to increase min_nodes. Every allocation has an overhead and question is how does min_modes of 4 compare to this overhead. I mean it's desirable that overhead/payload ratio is low.

Reply:
I see the problem now.  Is there such a thing as _code blindness_?
I've fixed this and moved the minimum test up a bit.

As for postponing the allocation of st->data, it is probably a good idea.
Likewise, `min_nodes` of 4 seems low to be but without some statistics it is impossible to tell.
I'd suggest a separate PR for each.

Reply:
I've created #4426 to track the suggestions for the future.